### ISSUE TITLE
How to adapt the code for BERT?

### ISSUE THREAD

**Issue by zhang373** [dholzmueller/bmdal_reg#5]
Hello, sorry for bothering you.
I am using "bmdal_reg" for a regression problem with BERT(https://arxiv.org/abs/1810.04805)
The error happen in this code:
`    def pop_feature_data(self) -> FeatureData:
        # remove the hooks
        self._input_hook.remove()
        self._grad_output_hook.remove()
        # compute the adjusted input \tilde{x} from the paper
        # TODO: Clean this!
        print("line 175 in layer_features.py: ",self.weight_factor, self._input_data.shape)
        print("line 176 in layer_features.py: ",self.bias_factor, torch.ones((self._input_data.shape[0], 1, 1), device=self._input_data.device).shape)
        print("line 177 in layer_features.py: ",self._grad_output_data.shape,"\n")
        fd=None
        inp = torch.cat([self.weight_factor * self._input_data,
                         self.bias_factor * torch.ones(self._input_data.shape[0], 1, device=self._input_data.device)],
                        dim=1)
        # feature data for the two IdentityFeatureMaps in the ProductFeatureMap, given by inputs and grad_outputs
        fd = ListFeatureData([TensorFeatureData(inp), TensorFeatureData(self._grad_output_data)])

        # allow to release memory earlier
        self._input_data = None
        self._grad_output_data = None
        
        return fd`

I don't why do we need to use: 
`        inp = torch.cat([self.weight_factor * self._input_data,
                         self.bias_factor * torch.ones(self._input_data.shape[0], 1, device=self._input_data.device)],
                        dim=1)`
Also, in the demo of bmdal, the input data shape is like [batchsize,2]. The output of cat process is like [100,3]
However, why I use BERT, the input data shape is like [batchsize, max_padding_length, word_vector](like[100,48,736]). How can I motify the code in bmdal?

**Comment by zhang373** [comment]
I adapt the code to 

```    def pop_feature_data(self) -> FeatureData:
        # remove the hooks
        self._input_hook.remove()
        self._grad_output_hook.remove()
        # compute the adjusted input \tilde{x} from the paper
        # TODO: Clean this!
        print("line 175 in layer_features.py: ",self.weight_factor, self._input_data.shape)
        print("line 176 in layer_features.py: ",self.bias_factor, torch.ones((self._input_data.shape[0],self._input_data.shape[1], 1), device=self._input_data.device).shape)
        print("line 177 in layer_features.py: ",self._grad_output_data.shape)
        fd=None
        inp = torch.cat([self.weight_factor * self._input_data,
                         self.bias_factor * torch.ones((self._input_data.shape[0],self._input_data.shape[1], 1), device=self._input_data.device)],
                        dim=2)
        print("line 182 in layer_features.py: ",inp.shape,"\n")
        # feature data for the two IdentityFeatureMaps in the ProductFeatureMap, given by inputs and grad_outputs
        fd = ListFeatureData([TensorFeatureData(inp), TensorFeatureData(self._grad_output_data)])

        # allow to release memory earlier
        self._input_data = None
        self._grad_output_data = None
        
        return fd
```

It works.


However, I still don't know why

Edit @dholzmueller: Adjusted code block

**Comment by dholzmueller** [comment]
The reasoning behind the `torch.cat()` part is in Equations (4) - (6) in the paper, the idea is that the `torch.ones()` part comes from the gradients wrt the biases, while the rest comes from the gradients wrt the weights. I don't know if your code is correct since I don't know exactly how the respective BERT layer works - is it like a vectorized linear layer? With the same weight matrix for the max_padding_length dimension?

**Comment by zhang373** [comment]
Thank you for your kind reply! I will check equation 4-6.
I don't know the BERT detailed as I didn't read the source code. But according to my knowledge gain from the paper(BERT). 
The vectorization process is done by word2vector before sending the data to bert. The layer here is just a liner layer process 3d data. The weight are shared.

By the way. bmdal is really a great work. Thank you for your kind contribution.

**Comment by dholzmueller** [comment]
Thanks. If you can find a specific equation / implementation of the 3D linear layer I can maybe comment on the situation.
